src/stream.rs

```diff
@@ -202,27 +229,128 @@ pub(crate) fn connect_host(unit: &Unit, hostname: &str, port: u16) -> Result<Tcp
         stream.set_write_timeout(None).ok();
     }
 
-    if let Some(ref proxy) = unit.proxy {
-        write!(stream, "{}", proxy.connect(hostname, port)).unwrap();
-        stream.flush()?;
+    if proto == Some(Proto::HTTPConnect) {
+        if let Some(ref proxy) = unit.proxy {
+            write!(stream, "{}", proxy.connect(hostname, port)).unwrap();
+            stream.flush()?;
 
-        let mut proxy_response = Vec::new();
+            let mut proxy_response = Vec::new();
 
-        loop {
-            let mut buf = vec![0; 256];
-            let total = stream.read(&mut buf)?;
-            proxy_response.append(&mut buf);
-            if total < 256 {
-                break;
+            loop {
+                let mut buf = vec![0; 256];
+                let total = stream.read(&mut buf)?;
+                proxy_response.append(&mut buf);
+                if total < 256 {
+                    break;
+                }
             }
+
+            Proxy::verify_response(&proxy_response)?;
         }
+    }
+
+    Ok(stream)
+}
 
-        crate::Proxy::verify_response(&proxy_response)?;
+#[cfg(feature = "socks-proxy")]
+fn connect_socks5(
+    proxy: Proxy,
+    timeout_connect: u64,
+    proxy_addr: SocketAddr,
+    hostname: String,
+    port: u16,
+) -> Result<TcpStream, std::io::Error> {
+    let host_addrs: Vec<SocketAddr> = format!("{}:{}", hostname, port)
+        .to_socket_addrs()
+        .map_err(|e| std::io::Error::new(ErrorKind::NotFound, format!("DNS failure: {}.", e)))?
+        .collect();
+
+    if host_addrs.is_empty() {
+        return Err(std::io::Error::new(
+            ErrorKind::NotFound,
+            format!("DNS lookup failed for {}:{}.", hostname, port),
+        ));
     }
 
+    let stream = if timeout_connect > 0 {
+        let master_signal = Arc::new((Mutex::new(false), Condvar::new()));
+        let slave_signal = master_signal.clone();
+        let (tx, rx) = channel();
+        let host_addr = host_addrs[0];
+        thread::spawn(move || {
+            let (lock, cvar) = &*slave_signal;
+            if tx
+                .send(get_socks5_stream(&proxy, &proxy_addr, &host_addr))
+                .is_ok()
+            {
+                let mut done = lock.lock().unwrap();
+                *done = true;
+                cvar.notify_one();
+            }
```

Comment:
Ouch, I understand why you're doing this, but it feels wrong to launch a thread. Maybe make a code comment about:

1. Why the need for a thread.
2. A 'TODO' about approaching the upstream crate so that `Socks5Stream::connect...` can take a timeout.
3. Why `tx.send(…).is_ok()` being false is an expected code path.

Reply:
Thanks for the quick response.
Please have a look at the 3 most recent commits and let me know if you'd like any further changes.

Reply:
BTW, I completely agree about it being wrong to launch a thread, and trying to fix it upstream is my intention. I have briefly looked at Socks5Stream and I can see it using a BufReader internally, so it wasn't immediately obvious to me how to change that hence my choice to implement something basic initially.